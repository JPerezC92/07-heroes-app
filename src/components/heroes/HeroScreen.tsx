import { useEffect, useMemo, useState } from "react";
import { Redirect, RouteComponentProps, useParams } from "react-router-dom";
import { dynamicImportImg } from "../../helper/dynamicImportImg";
import { getHeroById } from "../../selectors/getHeroById";

const HeroScreen = ({ history }: RouteComponentProps) => {
  const { heroeId = "" }: { heroeId: string } = useParams();
  const [heroImage, setHeroImage] = useState("");
  const hero = useMemo(() => getHeroById(heroeId), [heroeId]);

  const handleClick = () =>
    history.length > 2 ? history.goBack() : history.push("/");

  useEffect(() => {
    if (heroeId.length > 0) {
      dynamicImportImg(heroeId).then((img) => setHeroImage(() => img));
    }
  }, [heroeId]);

  if (!hero) return <Redirect to="/" />;

  return (
    <div className="row mt-5">
      <div className="col-4">
        <img
          className="img-thumbnail animate__animated animate__fadeInLeft"
          src={heroImage}
          alt={hero.superhero}
        />
      </div>

      <div className="col-8">
        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            <b>Alter ego:</b> {hero.alter_ego}
          </li>
          <li className="list-group-item">
            <b>Publisher:</b> {hero.publisher}
          </li>
          <li className="list-group-item">
            <b>First Appearance:</b> {hero.first_appearance}
          </li>
        </ul>

        <h5>Characters</h5>

        <p>{hero.characters}</p>

        <button className="btn btn-outline-info" onClick={handleClick}>
          Return
        </button>
      </div>
    </div>
  );
};

export default HeroScreen;
