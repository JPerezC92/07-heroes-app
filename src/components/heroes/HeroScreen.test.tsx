import { mount } from "enzyme";

import { MemoryRouter, Route } from "react-router-dom";
import { heroScreenPropsFixture } from "../../test/fixtures/heroScreenPropsFixture";
import HeroScreen from "./HeroScreen";

describe("Test on <HeroScreen />", () => {
  test("should render <Redirect /> component if isn't arguments on URL'", () => {
    const HeroScreenProps = {
      ...heroScreenPropsFixture,
      history: {
        ...heroScreenPropsFixture.history,
        length: 10,
        push: jest.fn(),
        goBack: jest.fn(),
      },
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero"]}>
        <HeroScreen {...HeroScreenProps} />
      </MemoryRouter>
    );

    expect(wrapper.find("Redirect").exists()).toBeTruthy();
  });

  test("should render a hero if param exists", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/dc-batman"]}>
        <Route path="/hero/:heroeId" component={HeroScreen} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.contains("Bruce Wayne")).toBeTruthy();
  });

  test("should redirect with push()", () => {
    const HeroScreenProps = {
      ...heroScreenPropsFixture,
      history: {
        ...heroScreenPropsFixture.history,
        length: 1,
        push: jest.fn(),
        goBack: jest.fn(),
      },
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/dc-batman"]}>
        <Route
          path="/hero/:heroeId"
          component={() => <HeroScreen {...HeroScreenProps} />}
        />
      </MemoryRouter>
    );

    wrapper.find("button").simulate("click");

    expect(HeroScreenProps.history.goBack).not.toHaveBeenCalled();
    expect(HeroScreenProps.history.push).toHaveBeenCalledWith("/");
  });

  test("should redirect with goBack()", () => {
    const HeroScreenProps = {
      ...heroScreenPropsFixture,
      history: {
        ...heroScreenPropsFixture.history,
        length: 10,
        push: jest.fn(),
        goBack: jest.fn(),
      },
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/dc-batman"]}>
        <Route
          path="/hero/:heroeId"
          component={() => <HeroScreen {...HeroScreenProps} />}
        />
      </MemoryRouter>
    );

    wrapper.find("button").simulate("click");

    expect(HeroScreenProps.history.goBack).toHaveBeenCalled();
    expect(HeroScreenProps.history.push).not.toHaveBeenCalled();
  });

  test("should render <Redirect /> component if hero does not exists", () => {
    const heroScreenProps = {
      ...heroScreenPropsFixture,
      history: {
        ...heroScreenPropsFixture.history,
        length: 10,
        push: jest.fn(),
        goBack: jest.fn(),
      },
    };

    const heroId = "dc-batman123wewq";
    const wrapper = mount(
      <MemoryRouter initialEntries={[`/hero/${heroId}`]}>
        <Route
          path="/hero/:heroId"
          component={() => <HeroScreen {...heroScreenProps} />}
        />
      </MemoryRouter>
    );

    expect(wrapper.text()).toBe("");
  });
});
