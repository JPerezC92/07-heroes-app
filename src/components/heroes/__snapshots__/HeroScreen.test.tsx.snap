// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <HeroScreen /> should render a hero if param exists 1`] = `
<div
  className="row mt-5"
>
  <div
    className="col-4"
  >
    <img
      alt="Batman"
      className="img-thumbnail animate__animated animate__fadeInLeft"
      src="../assets/heroes/dc-batman.jpg"
    />
  </div>
  <div
    className="col-8"
  >
    <ul
      className="list-group list-group-flush"
    >
      <li
        className="list-group-item"
      >
        <b>
          Alter ego:
        </b>
         
        Bruce Wayne
      </li>
      <li
        className="list-group-item"
      >
        <b>
          Publisher:
        </b>
         
        DC Comics
      </li>
      <li
        className="list-group-item"
      >
        <b>
          First Appearance:
        </b>
         
        Detective Comics #27
      </li>
    </ul>
    <h5>
      Characters
    </h5>
    <p>
      Bruce Wayne
    </p>
    <button
      className="btn btn-outline-info"
      onClick={[Function]}
    >
      Return
    </button>
  </div>
</div>
`;
